File: src/Parameters/Vars/Vector.cpp
#include <limits>
#include "Vector.h"
#include "math.h"

Vector::Vector() {}

Vector::~Vector() {}

int Vector::size() {
    return vars.size();
}

double Vector::operator[](int i) {
    if (size() <= i) return std::numeric_limits<double>::max();
    return vars[i];
}

Vector Vector::operator+(Vector &&v) {
    int count = size();
    Vector summ = *(this->clone());
    if (count == v.size()) {
        for (int i = 0; i < count; i++) {
            summ.setAt(i, summ[i] + v[i]);
        }
    }
    return summ;
}

Vector Vector::operator+=(Vector &&v) {
    return this->operator+(std::move(v));
}

Vector Vector::operator-(Vector &&v) {
    int count = size();
    Vector razn = *(this->clone());
    if (count == v.size()) {
        for (int i = 0; i < count; i++) {
            razn.setAt(i, razn[i] - v[i]);
        }
    }
    return razn;
}

Vector Vector::operator-=(Vector &&v) {
    this->operator-(std::move(v));
}

Vector Vector::operator*(double a) {
    Vector proz = *(this->clone());
    for (int i = 0; i < size(); ++i) {
        proz.setAt(i, proz[i] * a);
    }
    return proz;
}

Vector Vector::operator*=(double a) {
    return this->operator*(a);
}

Vector Vector::operator/(double a) {
    return this->operator*(1 / a);
}

Vector Vector::operator/=(double a) {
    return this->operator*(1 / a);
}

double Vector::operator*(Vector &&v) {
    int coutn = size();
    double summ = 0.0;
    if (coutn == v.size()) {
        for (int i = 0; i < coutn; ++i) {
            summ += vars[i] * v[i];
        }
    } else {
        summ = std::numeric_limits<double>::max();
    }
    return summ;
}

Vector *Vector::clone() {
    Vector *v = new Vector();
    for (int i = 0; i < size(); ++i) {
        v->push(vars[i]);
    }
    return v;
}

void Vector::push(double var) {
    vars.push_back(var);
}

void Vector::operator=(double var) {
    vars.clear();
    vars.push_back(var);
}

std::string *Vector::toString() {
    std::string *str = new std::string("Vector:");
    char number[15];

    for (int i = 0; i < size(); ++i) {
        sprintf(number, " %0.10f", vars[i]);
        (*str) += number;
    }
    return str;
}

double Vector::norma() {
    if (vars.size() == 0) return 0.0;
    if (vars.size() == 1) return vars[0];

    double norm = 0.0;
    for (int i = 0; i < size(); ++i) {
        norm += vars[0] * vars[0];
    }

    return sqrt(norm);
}

Vector *Vector::random(int n, int absMax) {
    Vector *v = new Vector();
    for (int i = 0; i < n; ++i) {
        v->push(rand() % ((absMax + 1) * 2) - absMax);
    }
    return v;
}

Vector *Vector::getZeroBeside(int n, int i) {
    Vector *v = new Vector();
    for (int j = 0; j < n; ++j) {
        if (j != i) {
            v->push(0);
        } else {
            v->push(1);
        }
    }
    return v;
}

void Vector::setAt(int i, double val) {
    if (i < size()) {
        vars[i] = val;
    }
}

File: Parametrs/Vars/Vector.cs
using System.Collections;
using System;

namespace MethodsOptimization.src.Parametrs.Vars
{
    class Vector: ICloneable
    {
        /// <summary>
        /// переменные в векторе
        /// </summary>
        ArrayList vars = new ArrayList();

        public Vector() { }

        public Vector(double[] vs)
        {
            foreach(double v in vs)
            {
                vars.Add(v);
            }
        }

        /// <summary>
        /// Сложение элементов векторов
        /// </summary>
        /// <param name="v1">Первый вектор</param>
        /// <param name="v2">Второй вектор</param>
        /// <returns>Новый вектор суммы</returns>
        public static Vector operator +(Vector v1, Vector v2)
        {
            Vector sum = new Vector();
            if (v1.Size != v2.Size)
            {
                return new Vars.Vector();
            }
            for (int i = 0; i < v1.Size; i++)
            {
                sum.Push(v1[i] + v2[i]);
            }
            return sum;
        }

        /// <summary>
        /// Разность элементов векторов
        /// </summary>
        /// <param name="v1">Первый вектор</param>
        /// <param name="v2">Второй вектор</param>
        /// <returns>Новый вектор разности</returns>
        public static Vector operator -(Vector v1, Vector v2)
        {
            Vector raz = new Vector();
            if (v1.Size != v2.Size)
            {
                return new Vars.Vector();
            }
            for (int i = 0; i < v1.Size; i++)
            {
                raz.Push(v1[i] - v2[i]);
            }
            return raz;
        }

        /// <summary>
        /// Умножение вектора на скаляр
        /// </summary>
        /// <param name="v">Вектор</param>
        /// <param name="a">Скаляр</param>
        /// <returns>Новый вектор произведения вектора на скаляр</returns>
        public static Vector operator *(Vector v, double a)
        {
            Vector proz = new Vector();
            for (int i = 0; i < v.Size; i++)
            {
                proz.Push(v[i] * a);
            }
            return proz;
        }

        /// <summary>
        /// Умножение вектора на скаляр
        /// </summary>
        /// <param name="v">Вектор</param>
        /// <param name="a">Скаляр</param>
        /// <returns>Новый вектор произведения вектора на скаляр</returns>
        public static Vector operator *(double a, Vector v)
        {
            return v * a;
        }

        /// <summary>
        /// Деление вектора на скаляр
        /// </summary>
        /// <param name="v">Вектор</param>
        /// <param name="a">Скаляр</param>
        /// <returns>Новый вектор произведения вектора на скаляр</returns>
        public static Vector operator /(Vector v, double a)
        {
            return v * (1 / a);
        }

        /// <summary>
        /// Скалярное произведение векторов
        /// </summary>
        /// <param name="v1">Первый вектор</param>
        /// <param name="v2">Второй вектор</param>
        /// <returns>Скалярное произведение</returns>
        public static double operator *(Vector v1, Vector v2)
        {
            if (v1.Size != v2.Size)
            {
                return 0.0;
            }
            double sum = 0.0;
            for (int i = 0; i < v1.Size; i++)
            {
                sum += v1[i] * v2[i];
            }
            return sum;
        }

        /// <summary>
        /// Добавление в конец вектора одного элемента
        /// </summary>
        /// <param name="var"></param>
        public void Push(double var)
        {
            vars.Add(var);
        }

        /// <summary>
        /// Добавление в конец вектора несколько элементов
        /// </summary>
        /// <param name="vars"></param>
        public void Push(double[] vars)
        {
            foreach (double var in vars)
            {
                this.vars.Add(var);
            }
        }

        /// <summary>
        /// удалить i-й элемент
        /// </summary>
        /// <param name="number"></param>
        public void Remove(int number)
        {
            vars.RemoveAt(number);
        }

        /// <summary>
        /// Размер вектора
        /// </summary>
        /// <returns>Размер вектора</returns>
        public int Size
        {
            get
            {
                return vars.Count;
            }
        }

        /// <summary>
        /// Получение элемента вектора
        /// </summary>
        /// <param name="number">Номер элемента</param>
        /// <returns>Если выход за пределы вектора, то макс NaN, иначе элемент вектора</returns>
        public double this [int number]
        {
            get
            {
                if (!checkIndex(number))
                {
                    return double.NaN;
                }
                return (double)vars[number];
            }
            set
            {
                if (checkIndex(number))
                {
                    vars[number] = value;
                }
            }
        }

        /// <summary>
        /// Клонирование вектора
        /// </summary>
        /// <returns>Новый копированый вектор</returns>
        public object Clone()
        {
            Vector clone = new Vector();
            foreach(double var in vars) {
                clone.Push(var);
            }
            return clone;
        }

        public void Clear()
        {
            vars.Clear();
        }

        /// <summary>
        /// К виду строки
        /// </summary>
        /// <returns>Строка</returns>
        new public string ToString()
        {
            string str = "Vector:\n";
            int i = 0;
            foreach(double v in vars)
            {
                str += "\tvector[" + i.ToString() + "] = " + v.ToString("G8") + ";\n";
                i++;
            }
            return str;
        }

        /// <summary>
        /// Норма веектора
        /// </summary>
        public double Norma
        {
            get
            {
                if (Size == 0) return 0.0;
                if (Size == 1) return (double)vars[0];

                return Math.Sqrt(this * this);
            }
        }

        /// <summary>
        /// Получение случайного вектора
        /// </summary>
        /// <param name="n">Размер</param>
        /// <param name="absMax">максимально по модулю значение</param>
        /// <returns>Случайный вектор</returns>
        public static Vector random(int n, int absMax)
        {
            Vector random = new Vector();
            if (n < 0) return random;
            Random gen = new Random();
            for (int i = 0; i < n; i++)
            {
                random.Push(gen.Next(-absMax, absMax));
            }
            return random;
        }

        /// <summary>
        /// Получение нулевого вектора кроме i-ого
        /// </summary>
        /// <param name="n">Размер</param>
        /// <param name="i">Номер не нулевого элемента</param>
        /// <returns>Новый вектор</returns>
        public static Vector GetZeroBeside(int n, int i)
        {
            Vector zero = new Vector();
            if (n < 0 || i >= n) return zero;
            for (int j = 0; j < n; j++)
            {
                if (j == i)
                {
                    zero.Push(1.0);
                } else
                {
                    zero.Push(0.0);
                }

            }
            return zero;
        }

        /// <summary>
        /// Инвертирование элементов
        /// </summary>
        /// <param name="v">Инвертируемый вектор</param>
        /// <returns>Инвертированый вектор</returns>
        public static Vector operator-(Vector v)
        {
            Vector v2 = new Vector();
            for (int i = 0; i < v.Size; i++)
            {
                v2.Push(-v[i]);
            }
            return v2;
        }

        /// <summary>
        /// Проверка номера элемента на соответствие интервалу вектора
        /// </summary>
        /// <param name="number">номер элемента</param>
        /// <returns></returns>
        public bool checkIndex(int number)
        {
            return (number >= 0 && number < Size);
        }

        /// <summary>
        /// первая переменная из вектора
        /// </summary>
        public double First
        {
            get
            {
                return this[0];
            }
        }

        /// <summary>
        /// последняя переменная из вектора
        /// </summary>
        public double Last
        {
            get
            {
                return this[Size - 1];
            }
        }
    }
}
